design/authenticators_api.md

```diff
@@ -0,0 +1,178 @@
+| Conjur Operator | A Conjur user with enhanced privileges (loads policies to root, configures authenticators, etc.) |
+| Conjur User     | A developer/app that needs to login to Conjur to retrieve secrets |
+
+## Current Mechanism
+
+Before we go over the suggestion for the new API, let's understand the flaws of
+the current mechanism. Currently, anyone can run an `/authenticators` request 
+and will get the following response:
+```
+{
+    "installed": [
+        "authn",
+        "authn-1",
+        "authn-2",
+        "authn-3"
+    ],
+    "configured": [
+        "authn",
+        "authn-1"
+    ],
+    "enabled": [
+        "authn",
+        "authn-1",
+        "authn-2"
+    ]
+}
+```
+
+Let's go over the sections in the response to understand them:
+
+1. installed: these authenticators are implemented in conjur code 
+(e.g they have an `Authn-somehting::Authenticator` class with a `valid?` method).
+
+2. configured: these authenticators have a webservice called 
+`conjur/authn-something/webservice`.
+
+3. enabled: these authenticators are whitelisted in the CONJUR_AUTHENTICATORS variable.
+
+Section 1 has important data but it shouldn't be here, but rather in the docs. It has
+nothing to do with the given Conjur instance and just describes the authenticators that
+are implemented in Conjur.
+
+Sections 2 & 3 combined can help a conjur user to know if he can use an authenticator
+but that's not entirely true. some authenticators need further configuration than 
+just configuring a webservice and enabling the authenticator in the ENV (for example, 
+authn-oidc needs to load variables). So while a user will get _some_ value using this API, 
+it won't fully let him know that he can use the authenticator, which is the purpose of this.
+
+## Feature Overview 
+
+### Use Cases
+
+This feature is divided into 2 personas:
+
+#### Conjur operator
+
+**As a** Conjur operator\
+**I'd like to** know which authenticators are configured correctly\
+**So that** I can complete the configuration properly if it's incomplete (or remove
+it if it's not needed)
+
+#### Conjur user
+
+**As a** Conjur user\
+**I'd like to** know which authenticators are available for authentication\
+**So that** I can authenticate with a properly configured one
+
+Note that both personas would like to know the same thing (configured correctly == available for authentication)
+but the action they need it for is different. 
+
+We can learn 2 things from the above:
+
+1. The Conjur user needs to authenticate according to the response. So this request must be applicable **without** a Conjur access token. 
+1. The Conjur operator needs to know which actions are needed for completing the configuration,
+which may consist sensitive data. So this request must be applicable **only with** a Conjur access token.
+
+This leads us to 2 API endpoints:
+
+1. `/authenticators` (for Conjur users)
+2. `/authenticators/health` (for Conjur operators)
```

Comment:
Second endpoint `/authenticators/health` does not follow general guidelines for RESTful applications and is inconsistent with our endpoint definition patterns. I think there is also no need for an aggregated authenticator health check.

Anyways, to keep this short, this would fit better with our codebase patterns:
1. `/authenticators`
1. `/authenticators/<authenticator_id>`, where the `<authenticator_id>` is the individual authenticator information/status can be retrieved. We could even do `/authenticators/<authenticator_id>/status` if only status is needed

Reply:
can you elaborate on the RESTful guidelines or refer me to some place? i'd like to better understand this.

why are you against the aggregated authenticator health check? let's say it will be `/authenticators/status`.

Reply:
https://restfulapi.net/resource-naming/

The general idea is that if you have an endpoint that aggregates results about a "resource", then it should end with `/<plural_resource_name>`.

If you are pulling information about a single resource, then the suffix should be `/<plural_resource_name>/<resource_id>` with an option for specialized pages: `/<plural_resource_name>/<resource_id>/<specific_info>`

Reply:
> why are you against the aggregated authenticator health check? let's say it will be /authenticators/status.

Aggregated list is fine in some cases but then you create tight coupling between each component. 

With aggregated status page, `/authenticators` page needs to know:
- How to list authenticators
- How to query their individual statuses
- How to unify their statuses into a single response

This approach also cannot be run in parallel

With individual resource page, `/authenticators` page needs to know:
- Just how to list authenticators

Each individual page can then have self-contained code on how to evaluate its status

Reply:
now that i think about it, it makes sense also UX-wise. We intend the operator to use this endpoint after he finished to configure a single authenticator. so if, for example, i just configured the `authn-oidc` authenticator, then i'd like to know the status/health of that authenticator. in the aggregated response the `authn-oidc` health is the signal, and the rest of the authenticators are just noise.

Thanks for this point :)